renku/core/dataset/providers/web.py

```diff
@@ -0,0 +1,196 @@
+from typing import TYPE_CHECKING, List, Tuple
+
+from renku.core import errors
+from renku.core.constant import CACHE
+from renku.core.dataset.providers.api import ProviderApi, ProviderPriority
+from renku.core.util import communication
+from renku.core.util.dataset import check_url
+from renku.core.util.urls import remove_credentials
+
+if TYPE_CHECKING:
+    from renku.core.dataset.providers.models import DatasetAddMetadata
+    from renku.core.management.client import LocalClient
+
+
+class WebProvider(ProviderApi):
+    """A provider for downloading data from web URLs."""
+
+    priority = ProviderPriority.LOWEST
+    name = "Web"
+
+    @staticmethod
+    def supports(uri: str) -> bool:
+        """Whether or not this provider supports a given URI."""
+        is_remote, is_git = check_url(uri)
+        return is_remote and not is_git
+
+    @staticmethod
+    def supports_add() -> bool:
+        """Whether this provider supports adding data to datasets."""
+        return True
+
+    @staticmethod
+    def add(
+        client: "LocalClient",
+        uri: str,
+        destination: Path,
+        *,
+        extract: bool = False,
+        filename: str = None,
+        multiple: bool = False,
+        **kwargs,
+    ) -> List["DatasetAddMetadata"]:
+        """Add files from a URI to a dataset."""
+        return download_file(
+            client=client, uri=uri, destination=destination, extract=extract, filename=filename, multiple=multiple
+        )
+
+
+def _ensure_dropbox(url):
+    """Ensure dropbox url is set for file download."""
+    if not isinstance(url, urllib.parse.ParseResult):
+        url = urllib.parse.urlparse(url)
+
+    query = url.query or ""
+    if "dl=0" in url.query:
+        query = query.replace("dl=0", "dl=1")
+    else:
+        query += "dl=1"
+
+    url = url._replace(query=query)
+    return url
+
+
+def _provider_check(url):
+    """Check additional provider related operations."""
+    from renku.core.util import requests
+
+    url = requests.get_redirect_url(url)
+    url = urllib.parse.urlparse(url)
+
+    if "dropbox.com" in url.netloc:
+        url = _ensure_dropbox(url)
+
+    return urllib.parse.urlunparse(url)
+
+
+def download_file(
+    client: "LocalClient",
+    uri: str,
+    destination: Path,
+    extract: bool = False,
+    filename: str = None,
+    multiple: bool = False,
+) -> List["DatasetAddMetadata"]:
+    """Download a file from a URI and return its metadata."""
+    from renku.core.dataset.providers.models import DatasetAddAction, DatasetAddMetadata
+    from renku.core.util import requests
+
+    uri = requests.get_redirect_url(uri)  # TODO: Check that this is not duplicate
+    uri = _provider_check(uri)
+
+    try:
+        start = time.time() * 1e3
+
+        tmp_root, paths = requests.download_file(
+            base_directory=client.renku_path / CACHE, url=uri, filename=filename, extract=extract
+        )
+
+        exec_time = (time.time() * 1e3 - start) // 1e3
```

Comment:
I feel just using `round()` or `floor()` instead of the `1e3` code is cleaner

Reply:
I've changed this rate-limit logic a bit.